### PR TITLE
FixedPointInt256 library

### PR DESCRIPTION
**Pull request by haythem96**
Int256 fixed point library.

### CODE REVIEW

File: contracts/libs/FixedPointInt256.sol [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+    int256 private constant SCALING_FACTOR = 1e18;
+
+    /**
+     * @notice convert an unsigned integer to signed integer
+     * @param a uint to convert into a signed integer.
+     * @return the converted signed integer.
+     */
+    function fromUint(uint256 a) internal pure returns (int256) {
+        require(a < uint256(-1), "FixedPointInt256: out of int range");
```
Comment (aparnakr) [inline contracts/libs/FixedPointInt256.sol:22]: also a max value restriction right?

File: contracts/libs/FixedPointInt256.sol [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+        return int256(a);
+    }
+
+    /**
+     * @notice convert a signed integer to unsigned integer
+     * @param a int to convert into an unsigned integer.
+     * @return the converted unsigned integer.
+     */
+    function fromInt(int256 a) internal pure returns (int256) {
```
Comment (aparnakr) [inline contracts/libs/FixedPointInt256.sol:32]:
```suggestion
    function fromInt(int256 a) internal pure returns (uint256) {
```

File: contracts/libs/FixedPointInt256.sol [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    function sub(int256 a, int256 b) internal pure returns (int256) {
+        return a.sub(b);
+    }
+
+    function mul(int256 a, int256 b) internal pure returns (int256) {
+        return (a.mul(b)).div(SCALING_FACTOR);
+    }
+
+    function div(int256 a, int256 b) internal pure returns (int256) {
+        return (a.div(b)).mul(SCALING_FACTOR);
```
Comment (aparnakr) [inline contracts/libs/FixedPointInt256.sol:53]: maybe you want to round these? See maker's math library for deets

Comment (haythem96) [reply contracts/libs/FixedPointInt256.sol:53]: cool

File: contracts/libs/FixedPointInt256.sol (hunk 1 above)
Comment (haythem96) [reply contracts/libs/FixedPointInt256.sol:22]: yeah.

File: contracts/libs/FixedPointInt256.sol [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    using SignedSafeMath for int256;
+
+    int256 private constant SCALING_FACTOR = 1e18;
+
+    /**
+     * @notice convert an unsigned integer to signed integer
+     * @param a uint to convert into a signed integer.
+     * @return the converted signed integer.
+     */
+    function fromUint(uint256 a) internal pure returns (int256) {
```
Comment (aparnakr) [inline contracts/libs/FixedPointInt256.sol:21]: can we name it uintToInt?

File: contracts/libs/FixedPointInt256.sol [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+        return int256(a);
+    }
+
+    /**
+     * @notice convert a signed integer to unsigned integer
+     * @param a int to convert into an unsigned integer.
+     * @return the converted unsigned integer.
+     */
+    function fromInt(int256 a) internal pure returns (uint256) {
```
Comment (aparnakr) [inline contracts/libs/FixedPointInt256.sol:32]:
```suggestion
    function IntToUint(int256 a) internal pure returns (uint256) {
```

File: contracts/libs/FixedPointInt256.sol [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        if (a < 0) {
+            return uint256(-a);
+        } else {
+            return uint256(a);
```
Comment (aparnakr) [inline contracts/libs/FixedPointInt256.sol:36]: does this need an overflow / underflow check?

File: test/fixedPointInt256.test.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+      const expectedInt = new BigNumber(5)
+
+      assert.equal(
+        (await lib.testFromUint(uint)).toNumber(),
+        expectedInt.toNumber(),
+        'conversion from uint to int mismatch',
+      )
+    })
+
+    it('It should revert converting an unsigned integer greater than uint256(-1) to signed integer', async () => {
```
Comment (aparnakr) [inline test/fixedPointInt256.test.ts:27]: does this also work for 2^128 + 1?

**Comment by haythem96** [comment]
> Looks good to me. Should we also have lte, gte functions?

yeah good idea, will add those also.

File: contracts/libs/FixedPointInt256.sol (hunk 6 above)
Comment (haythem96) [reply contracts/libs/FixedPointInt256.sol:36]: I don't think so, it is meant to solve this issue https://solidity.readthedocs.io/en/latest/types.html#explicit-conversions

File: test/fixedPointInt256.test.ts (hunk 7 above)
Comment (haythem96) [reply test/fixedPointInt256.test.ts:27]: why 2^128 ?!

File: contracts/libs/FixedPointInt256.sol (hunk 4 above)
Comment (antoncoding) [reply contracts/libs/FixedPointInt256.sol:21]: agree

File: contracts/libs/FixedPointInt256.sol [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+    /**
+     * @notice multiply two signed integer
+     * @dev rounds to zero if a*b < SCALING_FACTOR / 2
+     * @param a signed integer
+     * @param b signed integer
+     * @return mul of two signed integer
+     */
+    function mul(int256 a, int256 b) internal pure returns (int256) {
+        return (a.mul(b)).add(SCALING_FACTOR / 2) / SCALING_FACTOR;
```
Comment (antoncoding) [inline contracts/libs/FixedPointInt256.sol:68]:
```suggestion
        return (a.mul(b)).add(SCALING_FACTOR / 2).div(SCALING_FACTOR);
```

Any reason why not using safemath `.div` here?

Comment (haythem96) [reply contracts/libs/FixedPointInt256.sol:68]: It is the same, `SCALING_FACTOR` is always equal to 1e18.

File: test/fixedPointInt256.test.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+      assert.equal(
+        (await lib.testFromInt(int)).toNumber(),
+        expectedUint.toNumber(),
+        'conversion from int to uint mismatch',
+      )
+    })
+  })
+
+  describe('Test Addition', () => {
```
Comment (antoncoding) [inline test/fixedPointInt256.test.ts:45]: I think we should test more edge cases:
* max positive number + min negative number
* max postitive + max positive
* min negative + min negative

Comment (haythem96) [reply test/fixedPointInt256.test.ts:45]: for addition, I'm basically using SignedSafeMath from OZ, so those edge cases are already tested.

File: test/fixedPointInt256.test.ts (hunk 7 above)
Comment (haythem96) [reply test/fixedPointInt256.test.ts:27]: yes.

File: test/fixedPointInt256.test.ts [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    it('Should convert from signed integer to unsigned integer', async () => {
+      const int = new BigNumber(-3)
+      const expectedUint = new BigNumber(3)
+
+      assert.equal(
+        (await lib.testFromInt(int)).toNumber(),
+        expectedUint.toNumber(),
+        'conversion from int to uint mismatch',
+      )
+    })
```
Comment (antoncoding) [inline test/fixedPointInt256.test.ts:42]:
```suggestion
    })
    it('Should convert from positive signed integer to unsigned integer', async () => {
      const int = new BigNumber(3)
      const expectedUint = new BigNumber(3)

      assert.equal(
        (await lib.testFromInt(int)).toNumber(),
        expectedUint.toNumber(),
        'conversion from int to uint mismatch',
      )
    })
```
